lms/services/http.py

```diff
@@ -7,7 +7,7 @@ class HTTPService:
     """Send HTTP requests with `requests` and receive the responses."""
 
     # This is here mostly to let auto-spec know about it in the tests
-    session: Session = None
+    session: Session = None  # typing: ignore
```

Comment:
Removing the default makes the test fail, which is the reason to have this in the first place.

Reply:
I had a look through the Python docs for `create_autospec`, which suggests some alternative ways of handling this. I include the diff below to illustrate, but `# typing: ignore` is simpler.

```diff
diff --git a/lms/services/http.py b/lms/services/http.py
index 8d0e73fca..8c282b3c5 100644
--- a/lms/services/http.py
+++ b/lms/services/http.py
@@ -6,8 +6,7 @@ from lms.services.exceptions import ExternalRequestError
 class HTTPService:
     """Send HTTP requests with `requests` and receive the responses."""
 
-    # This is here mostly to let auto-spec know about it in the tests
-    session: Session = None  # typing: ignore
+    session: Session
     """The underlying requests Session."""
 
     def __init__(self):
@@ -15,7 +14,6 @@ class HTTPService:
         # requests and urllib3 connection pooling is used (which means that
         # underlying TCP connections are re-used when making multiple requests
         # to the same host, e.g. pagination).
-
         # See https://docs.python-requests.org/en/latest/user/advanced/#session-objects
         self.session = Session()
 
diff --git a/tests/unit/lms/services/jstor/service_test.py b/tests/unit/lms/services/jstor/service_test.py
index d5152d0c3..ea4d86ad1 100644
--- a/tests/unit/lms/services/jstor/service_test.py
+++ b/tests/unit/lms/services/jstor/service_test.py
@@ -5,7 +5,7 @@ from unittest.mock import sentinel
 import pytest
 
 from lms.services import ExternalRequestError
-from lms.services.jstor.service import ArticleNotFound, JSTORService
+from lms.services.jstor.service import ArticleNotFound, HTTPService, JSTORService
 from tests import factories
 
 API_URL = "http://jstore_api.example.com"
@@ -200,7 +200,12 @@ class TestJSTORService:
 
     @pytest.fixture(autouse=True)
     def HTTPService(self, patch):
-        return patch("lms.services.jstor.service.HTTPService")
+        class HTTPServiceForTest(HTTPService):
+            session = None
+
+        return patch(
+            "lms.services.jstor.service.HTTPService", autospec=HTTPServiceForTest
+        )
 
     @pytest.fixture(autouse=True)
     def JWTService(self, patch):
diff --git a/tests/unit/lms/services/vitalsource/_client_test.py b/tests/unit/lms/services/vitalsource/_client_test.py
index a0c43921e..9da609b36 100644
--- a/tests/unit/lms/services/vitalsource/_client_test.py
+++ b/tests/unit/lms/services/vitalsource/_client_test.py
@@ -7,6 +7,7 @@ from requests import Request
 from lms.services.exceptions import ExternalRequestError
 from lms.services.vitalsource._client import (
     BookNotFound,
+    HTTPService,
     VitalSourceClient,
     VitalSourceConfigurationError,
     _VSUserAuth,
@@ -253,9 +254,13 @@ class TestVitalSourceClient:
 
     @pytest.fixture(autouse=True)
     def http_service(self, patch):
-        HTTPService = patch("lms.services.vitalsource._client.HTTPService")
+        class HTTPServiceForTest(HTTPService):
+            session = None
 
-        return HTTPService.return_value
+        cls = patch(
+            "lms.services.vitalsource._client.HTTPService", autospec=HTTPServiceForTest
+        )
+        return cls.return_value
 
     @pytest.fixture
     def _VSUserAuth(self, patch):
diff --git a/tests/unit/services.py b/tests/unit/services.py
index 6e71bc381..a43513842 100644
--- a/tests/unit/services.py
+++ b/tests/unit/services.py
@@ -1,6 +1,7 @@
 from unittest import mock
 
 import pytest
+from requests import Session
 
 from lms.product.plugin.course_copy import (
     CourseCopyFilesHelper,
@@ -234,9 +235,11 @@ def h_api(mock_service):
 
 @pytest.fixture
 def http_service(mock_service):
-    http_service = mock_service(HTTPService, service_name="http")
-    http_service.request.return_value = factories.requests.Response()
+    class HTTPServiceForTest(HTTPService):
+        session = mock.create_autospec(Session, spec_set=True, instance=True)
 
+    http_service = mock_service(HTTPServiceForTest, service_name="http")
+    http_service.request.return_value = factories.requests.Response()
     return http_service
```

Reply:
Should we then define the type as `Session | None`?